# • DOMAIN: Smartphone, Electronics
# • CONTEXT: 
- India is the second largest market globally for smartphones after China. About 134 million smartphones were sold across India in the year 2017 and is estimated to increase to about 442 million in 2022. India ranked second in the average time spent on mobile web by smartphone users across Asia Pacific. The combination of very high sales volumes and the average smartphone consumer behaviour has made India a very attractive market for foreign vendors. As per Consumer behaviour, 97% of consumers turn to a search engine when they are buying a product vs. 15% who turn to social media. If a seller succeeds to publish smartphones based on user’s behaviour/choice at the right place, there are 90% chances that user will enquire for the same. This Case Study is targeted to build a recommendation system based on individual consumer’s behaviour or choice.
# • DATA DESCRIPTION:
-  author : name of the person who gave the rating
- country : country the person who gave the rating belongs to
- data : date of the rating
- domain: website from which the rating was taken from
- extract: rating content
- language: language in which the rating was given
- product: name of the product/mobile phone for which the rating was given
- score: average rating for the phone
- score_max: highest rating given for the phone
- source: source from where the rating was taken
# PROJECT OBJECTIVE:
- We will build a recommendation system using popularity based and collaborative filtering methods to recommend mobile phones to a user which are most popular and personalised respectively.
# Steps and tasks: 
# 1. Import the necessary libraries and read the provided CSVs as a data frame and perform the below steps. 
- A. Merge all the provided CSVs into one dataFrame. 








In [1]:
import pandas as pd
import numpy as np

In [2]:
df1 = pd.read_csv('phone_user_review_file_1.csv',encoding='latin-1')
print('-> phone_user_review_file_1.csv')
print(df1.head())
print("*"*80)
df2 = pd.read_csv('phone_user_review_file_2.csv',encoding='latin-1')
print('-> phone_user_review_file_2.csv')
print(df2.head())
print("*"*80)
df3 = pd.read_csv('phone_user_review_file_3.csv',encoding='latin-1')
print('-> phone_user_review_file_3.csv')
print(df3.head())
print("*"*80)
df4 = pd.read_csv('phone_user_review_file_4.csv',encoding='latin-1')
print('-> phone_user_review_file_4.csv')
print(df4.head())
print("*"*80)
df5 = pd.read_csv('phone_user_review_file_5.csv',encoding='latin-1')
print('-> phone_user_review_file_5.csv')
print(df5.head())
print("*"*80)
df6 = pd.read_csv('phone_user_review_file_6.csv',encoding='latin-1')
print('-> phone_user_review_file_6.csv')
print(df6.head())

-> phone_user_review_file_1.csv
                        phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nic

In [3]:
df1.shape,df2.shape,df3.shape,df4.shape,df5.shape,df6.shape

((374910, 11),
 (114925, 11),
 (312961, 11),
 (98284, 11),
 (350216, 11),
 (163837, 11))

In [4]:
df = pd.concat([df1,df2,df3,df4,df5,df6], axis=0)
df.shape

(1415133, 11)

In [5]:
print(len(df1.index)+len(df2.index)+len(df3.index)+len(df4.index)+len(df5.index)+len(df6.index))

1415133


- B. Explore, understand the Data and share at least 2 observations. 

In [6]:
df.sample(n=10)

phone_url        date lang country  \
68756                /cellphones/lg-google-nexus-5/    3/9/2016   en      ca   
18975                /cellphones/htc-touch-diamond/   6/20/2009   es      es   
149008                 /cellphones/huawei-honor-6x/   3/12/2017   en      in   
329909                       /cellphones/nokia-n97/    8/2/2009   de      de   
55448              /cellphones/lenovo-vibe-k4-note/    2/6/2016   en      in   
120717  /cellphones/samsung-galaxy-note-5-sm-n920f/  10/15/2015   en      us   
10419                       /cellphones/xiaomi-mi4/  12/18/2014   en      in   
70796                        /cellphones/oneplus-3/   7/18/2016   en      in   
13841           /cellphones/samsung-galaxy-s7-edge/   3/24/2016   en      us   
100098         /cellphones/nokia-9500-communicator/   2/14/2006   de      de   

           source        domain  score  score_max  \
68756      Amazon     amazon.ca   10.0       10.0   
18975   ShopMania  shopmania.es    8.0       10.0   
149008     Amazon     amazon.in   10.0       10.0   
329909     Amazon     amazon.de   10.0       10.0   
55448      Amazon     amazon.in   10.0       10.0   
120717     Amazon    amazon.com    2.0       10.0   
10419      Amazon     amazon.in   10.0       10.0   
70796      Amazon     amazon.in    6.0       10.0   
13841     Samsung   samsung.com    9.6       10.0   
100098       Ciao       ciao.de   10.0       10.0   

                                                  extract              author  \
68756   a bit of a wait but works just great phone is ...     Amazon Customer   
18975   Excelente Smartphone Pros: - GPS - WINDOWS - W...            CIVALDEZ   
149008                      Good phone . Camera excellent     Amazon Customer   
329909  Ich habe das Handy jetzt ca. 1 Woche und bin e...         R. Grzegorz   
55448                          It's very nice n cute ....     Amazon Customer   
120717  Horrible the phone keeps saying that the sim c...                 ana   
10419              It's an amazing, Fantastic smartphone.     Amazon Customer   
70796   Overall good but missing video calling feature...  VIRENDER MITHARWAL   
13841   I upgraded from a Note 4 to the S7 Edge. I hav...           Vanessa2    
100098  Endlich! Seit drei Tagen habe ich ihn! Die 100...             Treason   

                                                  product  
68756   LG ZTHY Google Nexus 5 LG D820 D821 2300mAh Ba...  
18975                                   HTC Touch Diamond  
149008                            Honor 6X (Silver, 32GB)  
329909  Microsoft Nokia N97 Smartphone (QWERTZ-Tastatu...  
55448                   Lenovo Vibe K4 Note (Black, 16GB)  
120717  Samsung N920 Unlocked Galaxy Note 5, GSM 32GB ...  
10419                        Mi Xiaomi Mi 4 (White, 16GB)  
70796                        OnePlus 3 (Soft Gold, 64 GB)  
13841                  Samsung Galaxy S7 edge 32GB (AT&T)  
100098                                         Nokia 9500

In [7]:
df.dtypes

phone_url     object
date          object
lang          object
country       object
source        object
domain        object
score        float64
score_max    float64
extract       object
author        object
product       object
dtype: object

In [8]:
df.describe().T

count      mean       std   min   25%   50%   75%   max
score      1351644.0   8.00706  2.616121   0.2   7.2   9.2  10.0  10.0
score_max  1351644.0  10.00000  0.000000  10.0  10.0  10.0  10.0  10.0

In [9]:
df.cov()

score  score_max
score      6.84409        0.0
score_max  0.00000        0.0

In [67]:
product = df['product']
product.unique()

array(['Samsung Galaxy S8',
       'Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unlocked (GSM + CDMA, US WARRANTY) (Orchid Gray)',
       'Samsung Galaxy S8 64GB (AT&T)', ..., 'Maxon MX-3204',
       'Maxon 3204', 'Alcatel Club Plus Handy'], dtype=object)

In [70]:
df.loc[df.domain == 'amazon.ca']

phone_url       date lang country  \
1545    /cellphones/samsung-galaxy-s6-edgeplus/   2/3/2016   en      ca   
9174        /cellphones/samsung-galaxy-s7-edge/   5/7/2016   en      ca   
9175        /cellphones/samsung-galaxy-s7-edge/   5/7/2016   en      ca   
9176        /cellphones/samsung-galaxy-s7-edge/   5/7/2016   en      ca   
9177        /cellphones/samsung-galaxy-s7-edge/   5/7/2016   en      ca   
...                                         ...        ...  ...     ...   
72611            /cellphones/samsung-sgh-s730i/  8/13/2013   en      ca   
72615            /cellphones/samsung-sgh-s730i/   7/3/2013   en      ca   
72618            /cellphones/samsung-sgh-s730i/  5/22/2013   en      ca   
72620            /cellphones/samsung-sgh-s730i/  4/17/2013   en      ca   
108452            /cellphones/motorola-razr-v3/  8/10/2015   en      ca   

        source     domain  score  score_max  \
1545    Amazon  amazon.ca   10.0       10.0   
9174    Amazon  amazon.ca   10.0       10.0   
9175    Amazon  amazon.ca    6.0       10.0   
9176    Amazon  amazon.ca    8.0       10.0   
9177    Amazon  amazon.ca    2.0       10.0   
...        ...        ...    ...        ...   
72611   Amazon  amazon.ca    4.0       10.0   
72615   Amazon  amazon.ca    4.0       10.0   
72618   Amazon  amazon.ca    8.0       10.0   
72620   Amazon  amazon.ca    8.0       10.0   
108452  Amazon  amazon.ca    2.0       10.0   

                                                  extract           author  \
1545    L'écran est superbe. Les couleurs très vive ...            Denis   
9174                       Was new and in good condition.              Raj   
9175    great item,does exacly what it needs to not ba...  Amazon Customer   
9176    Good phone, takes nice pictures and video. Sam...  Amazon Customer   
9177    Although this phone is sold as an unlocked pho...  Amazon Customer   
...                                                   ...              ...   
72611   Let me start with the positive: Easy to naviga...    Laura Webster   
72615   Phone works very well, very impressed with the...    Jeremy Aucoin   
72618   I would recommend this phone to anyone that wa...         RAgathon   
72620   this mobile phone is very esy to use and very ...     Assamoi Assi   
108452  This phone is not a US version, it has arrabic...          Td Visa   

                                                  product  
1545    Samsung Galaxy S6 EDGE PLUS 32GB Black (SM-G92...  
9174    Samsung Galaxy S7 Edge Factory Unlocked Phone ...  
9175    Samsung Galaxy S7 Edge Factory Unlocked Phone ...  
9176    Samsung Galaxy S7 Edge Factory Unlocked Phone ...  
9177    Samsung Galaxy S7 Edge Factory Unlocked Phone ...  
...                                                   ...  
72611   Samsung UNLOCKED Samsung Galaxy Discover SGH-S...  
72615   Samsung UNLOCKED Samsung Galaxy Discover SGH-S...  
72618   Samsung UNLOCKED Samsung Galaxy Discover SGH-S...  
72620   Samsung UNLOCKED Samsung Galaxy Discover SGH-S...  
108452  Motorola RAZR V3 Unlocked Phone with Camera an...  

[8055 rows x 11 columns]

- From the above data it is clear that it has more amount of text data and many of the features are irrelvant in developing a recommendation system. Also a few insights such as five point summary, mean, avg and correlation of the numerical features are determined.

- C. Round off scores to the nearest integers. 

In [10]:
df_round = df
df_round

phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain  score  score_max  \
0       Verizon Wireless  verizonwireless.com   10.0       10.0   
1            Phone Arena       phonearena.com   10.0       10.0   
2                 Amazon           amazon.com    6.0       10.0   
3                Samsung          samsung.com    9.2       10.0   
4       Verizon Wireless  verizonwireless.com    4.0       10.0   
...                  ...                  ...    ...        ...   
163832              Ciao              ciao.de    2.0       10.0   
163833              Ciao              ciao.de   10.0       10.0   
163834              Ciao              ciao.de    2.0       10.0   
163835              Ciao              ciao.de    8.0       10.0   
163836              Ciao              ciao.de    2.0       10.0   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1415133 rows x 11 columns]

In [11]:
df_round.score = df_round.score.round()
df_round

phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain  score  score_max  \
0       Verizon Wireless  verizonwireless.com   10.0       10.0   
1            Phone Arena       phonearena.com   10.0       10.0   
2                 Amazon           amazon.com    6.0       10.0   
3                Samsung          samsung.com    9.0       10.0   
4       Verizon Wireless  verizonwireless.com    4.0       10.0   
...                  ...                  ...    ...        ...   
163832              Ciao              ciao.de    2.0       10.0   
163833              Ciao              ciao.de   10.0       10.0   
163834              Ciao              ciao.de    2.0       10.0   
163835              Ciao              ciao.de    8.0       10.0   
163836              Ciao              ciao.de    2.0       10.0   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1415133 rows x 11 columns]

- D. Check for missing values. Impute the missing values, if any. 

In [12]:
df_round.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
dtype: int64

In [13]:
df_round = df_round.dropna(subset=['author'])
df_round.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        60893
score_max    60893
extract      15515
author           0
product          0
dtype: int64

In [14]:
df_n = df_round
df_n.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        60893
score_max    60893
extract      15515
author           0
product          0
dtype: int64

In [15]:
df_n["score_max"].fillna("10", inplace = True)
df_n

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain  score score_max  \
0       Verizon Wireless  verizonwireless.com   10.0        10   
1            Phone Arena       phonearena.com   10.0        10   
2                 Amazon           amazon.com    6.0        10   
3                Samsung          samsung.com    9.0        10   
4       Verizon Wireless  verizonwireless.com    4.0        10   
...                  ...                  ...    ...       ...   
163832              Ciao              ciao.de    2.0        10   
163833              Ciao              ciao.de   10.0        10   
163834              Ciao              ciao.de    2.0        10   
163835              Ciao              ciao.de    8.0        10   
163836              Ciao              ciao.de    2.0        10   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1351931 rows x 11 columns]

In [16]:
df_n.dtypes

phone_url     object
date          object
lang          object
country       object
source        object
domain        object
score        float64
score_max     object
extract       object
author        object
product       object
dtype: object

In [17]:
df_n.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        60893
score_max        0
extract      15515
author           0
product          0
dtype: int64

In [18]:
df_mean = df_n["score"].mean()

In [19]:
df_mean

8.011768824775103

In [20]:
df_n["score"].fillna("8", inplace = True)
df_n

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain score score_max  \
0       Verizon Wireless  verizonwireless.com    10        10   
1            Phone Arena       phonearena.com    10        10   
2                 Amazon           amazon.com     6        10   
3                Samsung          samsung.com     9        10   
4       Verizon Wireless  verizonwireless.com     4        10   
...                  ...                  ...   ...       ...   
163832              Ciao              ciao.de     2        10   
163833              Ciao              ciao.de    10        10   
163834              Ciao              ciao.de     2        10   
163835              Ciao              ciao.de     8        10   
163836              Ciao              ciao.de     2        10   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1351931 rows x 11 columns]

In [21]:
df_n.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score            0
score_max        0
extract      15515
author           0
product          0
dtype: int64

- E. Check for duplicate values and remove them, if any. 

In [22]:
df_n.duplicated().sum()

5049

In [23]:
df_n.drop_duplicates()

phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain score score_max  \
0       Verizon Wireless  verizonwireless.com    10        10   
1            Phone Arena       phonearena.com    10        10   
2                 Amazon           amazon.com     6        10   
3                Samsung          samsung.com     9        10   
4       Verizon Wireless  verizonwireless.com     4        10   
...                  ...                  ...   ...       ...   
163832              Ciao              ciao.de     2        10   
163833              Ciao              ciao.de    10        10   
163834              Ciao              ciao.de     2        10   
163835              Ciao              ciao.de     8        10   
163836              Ciao              ciao.de     2        10   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1346882 rows x 11 columns]

- F. Keep only 1 Million data samples. Use random state=612. 

In [24]:
df_new = df_n.sample(n=1000000, random_state=612)
df_new

phone_url        date lang country  \
101825  /cellphones/sony-xperia-z1-compact-d5503/   3/28/2014   fr      fr   
174506                    /cellphones/nokia-n900/    6/4/2010   pt      br   
162769                  /cellphones/asus-padfone/  12/10/2012   es      es   
211419           /cellphones/blackberry-passport/  12/22/2014   en      gb   
42030                       /cellphones/nokia-n9/  16/11/2012   ru      ru   
...                                           ...         ...  ...     ...   
61963               /cellphones/huawei-ascend-g6/   10/5/2014   de      de   
108249                    /cellphones/nokia-6101/   6/10/2007   pt      br   
157864                      /cellphones/nokia-c6/   4/27/2011   ru      ru   
104740                 /cellphones/motorola-v635/   3/19/2005   de      de   
175238                /cellphones/huawei-p8-lite/  10/20/2016   it      it   

              source                       domain score score_max  \
101825        Amazon                    amazon.fr     4        10   
174506  MercadoLibre  opiniao.mercadolivre.com.br     8        10   
162769          Ciao                      ciao.es     8        10   
211419        Amazon                 amazon.co.uk    10        10   
42030     Irecommend                irecommend.ru    10        10   
...              ...                          ...   ...       ...   
61963         Amazon                    amazon.de    10        10   
108249  MercadoLibre  opiniao.mercadolivre.com.br    10        10   
157864        Yandex             market.yandex.ru     2        10   
104740  Inside-handy              inside-handy.de     7        10   
175238        Amazon                    amazon.it     8        10   

                                                  extract           author  \
101825  Je suis contente de cette téléphone; les éq...    BUI THANH TRA   
174506  UM BOM APARELHO PEQUENO COMPUTADORZINHO DE MAO...   DOURIVAL FILHO   
162769  ASUS PadFone A66 Esto sí que es una novedad...          nfanjul   
211419  A sophisticated, excellent product. All of the...         Londoner   
42030                                                 NaN   ?????€???? 538   
...                                                   ...              ...   
61963   Das Huawei ist sehr handlich, die Kamara ist s...  Marcus Kowatsch   
108249  Produto de muito boa qualidade atende as neces...         GISALFHA   
157864  Ну не знаю, разочарована Ð...       yahochu107   
104740  Da kommt es also daher, aus einer kleinen schm...     handballfoto   
175238  Smartphone molto bello e maneggevole, con molt...   Marcello flash   

                                                  product  
101825  Sony Mobile Xperia Z1 Smartphone débloqué (4...  
174506                                         Nokia N900  
162769                                   ASUS PadFone A66  
211419  BlackBerry Passport 4.5-Inch SIM-Free Smartpho...  
42030                                            Nokia N9  
...                                                   ...  
61963   Huawei Ascend G6 Smartphone (11,4 cm (4,5 Zoll...  
108249                                         Nokia 6101  
157864                                        Nokia C6-01  
104740                                      Motorola V635  
175238  Huawei P8 Lite Smartphone, Display 5" IPS, Pro...  

[1000000 rows x 11 columns]

- G. Drop irrelevant features. Keep features like Author, Product, and Score. 

In [25]:
df_new1 = df_new.drop(['phone_url', 'date','lang','country','source','domain','score_max','extract'], axis=1)

# 2. Answer the following questions.
- A. Identify the most rated products. 



In [26]:
#print(df_new.groupby('product').size().sort_values(ascending=False))
df_top = df_new1
df_top1 = df_top.drop(['score','author'],axis=1)
df_top1['count'] = df_top1.groupby('product')['product'].transform('count')
df_top1.sort_values(by='count',ascending=False)


product  count
49564                    Lenovo Vibe K4 Note (White,16GB)   3866
48991                    Lenovo Vibe K4 Note (White,16GB)   3866
56256                    Lenovo Vibe K4 Note (White,16GB)   3866
50637                    Lenovo Vibe K4 Note (White,16GB)   3866
53331                    Lenovo Vibe K4 Note (White,16GB)   3866
...                                                   ...    ...
63067                 BlackBerry 9900 Bold black unlocked      1
81466   Doogee DG580 - Smartphone libre Android (panta...      1
153648                                             A1228c      1
210043                                  Sony Ericsson TXT      1
10736                                    Xiaomi (27) Mi 4      1

[1000000 rows x 2 columns]

- B. Identify the users with most number of reviews.

In [27]:
#print(df_new.groupby('author').size().sort_values(ascending=False))
df_topn = df_new1
df_topn1 = df_topn.drop(['score','product'],axis=1)
df_topn1['count'] = df_topn1.groupby('author')['author'].transform('count')
df_topn1.sort_values(by='count',ascending=False)

author  count
71816   Amazon Customer  56846
17591   Amazon Customer  56846
243646  Amazon Customer  56846
92873   Amazon Customer  56846
98113   Amazon Customer  56846
...                 ...    ...
120880     Bassam Buali      1
155100           arne20      1
176155        komix1998      1
48943            geha97      1
9746     Charles N Webb      1

[1000000 rows x 2 columns]

In [28]:
df_new1['p_count'] = df_new1['product'].map(df_new1['product'].value_counts())
#df_new1.sort_values(by='p_count',ascending=False)
df_new1['a_count'] = df_new1['author'].map(df_new1['author'].value_counts())
df_new1

score           author  \
101825     4    BUI THANH TRA   
174506     8   DOURIVAL FILHO   
162769     8          nfanjul   
211419    10         Londoner   
42030     10   ?????€???? 538   
...      ...              ...   
61963     10  Marcus Kowatsch   
108249    10         GISALFHA   
157864     2       yahochu107   
104740     7     handballfoto   
175238     8   Marcello flash   

                                                  product  p_count  a_count  
101825  Sony Mobile Xperia Z1 Smartphone débloqué (4...       15        1  
174506                                         Nokia N900      208        1  
162769                                   ASUS PadFone A66        2       18  
211419  BlackBerry Passport 4.5-Inch SIM-Free Smartpho...      104        5  
42030                                            Nokia N9      129        2  
...                                                   ...      ...      ...  
61963   Huawei Ascend G6 Smartphone (11,4 cm (4,5 Zoll...       43        1  
108249                                         Nokia 6101      109        1  
157864                                        Nokia C6-01      269        1  
104740                                      Motorola V635       40        1  
175238  Huawei P8 Lite Smartphone, Display 5" IPS, Pro...     1149       42  

[1000000 rows x 5 columns]

- C. Select the data with products having more than 50 ratings and users who have given more than 50 ratings. Report the shape of the final dataset. 

In [29]:
df_newu = df_new1[(df_new1['p_count'] > 50) & (df_new1['a_count'] > 50)]
df_newu

score           author  \
170533    10             John   
66599     10           Sergio   
41846     10             Nick   
33065     10            e-bit   
185093     8  Client d'Amazon   
...      ...              ...   
21251      2  Amazon Customer   
139189    10            e-bit   
75906     10  Amazon Customer   
69531      2  Amazon Customer   
141300     4  Amazon Customer   

                                                  product  p_count  a_count  
170533  Apple iPhone 3GS 8GB Black Factory Unlocked / ...       72      373  
66599   LG G2 - Smartphone libre Android (pantalla 5.2...      188      243  
41846   Sim Free Apple iPhone 5S 16GB Mobile Phone - S...      260      191  
33065           Smartphone Samsung Galaxy S7 SM-G930 32GB       89     6371  
185093  Huawei - P8 Lite - Smartphone Débloqué - 4G ...      134     5730  
...                                                   ...      ...      ...  
21251                 Lenovo Vibe K5 (Grey, VoLTE update)     1560    56846  
139189                  Smartphone Asus ZenFone 3 ZE520KL       92     6371  
75906                         OnePlus 3 (Graphite, 64 GB)     2993    56846  
69531                  YU Yuphoria YU5010A (Black+Silver)     1161    56846  
141300  Motorola Moto Z Play with Style Mod (Black, 32GB)      392    56846  

[108174 rows x 5 columns]

- The final shape of the dataset contains 5 columns and 108174 rows

In [30]:
df_newu.dtypes

score      object
author     object
product    object
p_count     int64
a_count     int64
dtype: object

In [31]:
df_newu.describe()

p_count        a_count
count  108174.000000  108174.000000
mean      533.678980   24066.111737
std       822.064001   26678.816329
min        51.000000      51.000000
25%        86.000000     210.000000
50%       184.000000    6371.000000
75%       547.000000   56846.000000
max      3866.000000   56846.000000

In [32]:
df_pbr = df_newu

# 3. Build a popularity based model and recommend top 5 mobile phones. 

In [33]:
df_pbr["score"] = df_pbr["score"].astype(object).astype(float)

<ipython-input-33-83736c0a68c2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_pbr["score"] = df_pbr["score"].astype(object).astype(float)


In [34]:
df_pbr.groupby('product')['score'].mean().sort_values(ascending=False).head()  

product
3100 / 3120                                                                                                                                            10.0
Samsung Galaxy S3 mini I8190 Smartphone (10,2 cm (4 Zoll) AMOLED Display, Dual-Core, 1GHz, 1GB RAM, 5 Megapixel Kamera, Android 4.1) sapphire-black    10.0
LG KF700                                                                                                                                               10.0
LG KP110                                                                                                                                               10.0
Samsung Galaxy S3                                                                                                                                      10.0
Name: score, dtype: float64

In [35]:
df_pbr.groupby('product')['score'].count().sort_values(ascending=False).head()  

product
Lenovo Vibe K4 Note (White,16GB)       2299
Lenovo Vibe K4 Note (Black, 16GB)      1873
OnePlus 3 (Graphite, 64 GB)            1365
OnePlus 3 (Soft Gold, 64 GB)           1326
Lenovo Vibe K5 (Gold, VoLTE update)    1195
Name: score, dtype: int64

In [36]:
rm_count = pd.DataFrame(df_pbr.groupby('product')['score'].mean()) 

In [37]:
rm_count['score_counts'] = pd.DataFrame(df_pbr.groupby('product')['score'].count())  

- Top 5 phone recommendation based on count of reviews and the score

In [38]:
rm_count.sort_values(by='score_counts',ascending=False).head()

score  score_counts
product                                                    
Lenovo Vibe K4 Note (White,16GB)     6.988256          2299
Lenovo Vibe K4 Note (Black, 16GB)    7.082755          1873
OnePlus 3 (Graphite, 64 GB)          8.493773          1365
OnePlus 3 (Soft Gold, 64 GB)         8.289593          1326
Lenovo Vibe K5 (Gold, VoLTE update)  6.634310          1195

In [39]:
rm_count.sort_values(by='score',ascending=False).head()

score  score_counts
product                                                                
3100 / 3120                                          10.0             1
Nokia Lumia 735 Smartphone débloqué 4G (Ecran...   10.0             3
Apple iPhone 4S 16GB                                 10.0             1
Nokia 5250                                           10.0             3
Samsung Guru E1200                                   10.0             3

In [40]:
print(rm_count.max())
print(rm_count.min())

score             10.0
score_counts    2299.0
dtype: float64
score           1.0
score_counts    1.0
dtype: float64


# 4. Build a collaborative filtering model using SVD. You can use SVD from surprise or build it from scratch(Note: Incase you’re building it from scratch you can limit your data points to 5000 samples if you face memory issues). Build a collaborative filtering model using kNNWithMeans from surprise. You can try both user-based and item-based model. 

In [41]:
df_c = df_newu
df_c

score           author  \
170533   10.0             John   
66599    10.0           Sergio   
41846    10.0             Nick   
33065    10.0            e-bit   
185093    8.0  Client d'Amazon   
...       ...              ...   
21251     2.0  Amazon Customer   
139189   10.0            e-bit   
75906    10.0  Amazon Customer   
69531     2.0  Amazon Customer   
141300    4.0  Amazon Customer   

                                                  product  p_count  a_count  
170533  Apple iPhone 3GS 8GB Black Factory Unlocked / ...       72      373  
66599   LG G2 - Smartphone libre Android (pantalla 5.2...      188      243  
41846   Sim Free Apple iPhone 5S 16GB Mobile Phone - S...      260      191  
33065           Smartphone Samsung Galaxy S7 SM-G930 32GB       89     6371  
185093  Huawei - P8 Lite - Smartphone Débloqué - 4G ...      134     5730  
...                                                   ...      ...      ...  
21251                 Lenovo Vibe K5 (Grey, VoLTE update)     1560    56846  
139189                  Smartphone Asus ZenFone 3 ZE520KL       92     6371  
75906                         OnePlus 3 (Graphite, 64 GB)     2993    56846  
69531                  YU Yuphoria YU5010A (Black+Silver)     1161    56846  
141300  Motorola Moto Z Play with Style Mod (Black, 32GB)      392    56846  

[108174 rows x 5 columns]

In [42]:
df_c1 = df_c.drop(['p_count','a_count'],axis=1)
df_c1

score           author  \
170533   10.0             John   
66599    10.0           Sergio   
41846    10.0             Nick   
33065    10.0            e-bit   
185093    8.0  Client d'Amazon   
...       ...              ...   
21251     2.0  Amazon Customer   
139189   10.0            e-bit   
75906    10.0  Amazon Customer   
69531     2.0  Amazon Customer   
141300    4.0  Amazon Customer   

                                                  product  
170533  Apple iPhone 3GS 8GB Black Factory Unlocked / ...  
66599   LG G2 - Smartphone libre Android (pantalla 5.2...  
41846   Sim Free Apple iPhone 5S 16GB Mobile Phone - S...  
33065           Smartphone Samsung Galaxy S7 SM-G930 32GB  
185093  Huawei - P8 Lite - Smartphone Débloqué - 4G ...  
...                                                   ...  
21251                 Lenovo Vibe K5 (Grey, VoLTE update)  
139189                  Smartphone Asus ZenFone 3 ZE520KL  
75906                         OnePlus 3 (Graphite, 64 GB)  
69531                  YU Yuphoria YU5010A (Black+Silver)  
141300  Motorola Moto Z Play with Style Mod (Black, 32GB)  

[108174 rows x 3 columns]

In [43]:
df_c1.shape

(108174, 3)

In [44]:
df_c1['author'].unique()

array(['John', 'Sergio', 'Nick', 'e-bit', "Client d'Amazon",
       'Amazon Customer', 'Paul', 'Amazon Kunde', 'Cliente Amazon',
       'Robert', 'Ð\x9cаÑ\x80иÑ\x8f', 'Massimiliano', 'Peter', 'Nicole',
       'Bill', '?????????????????\x80', 'Ivan', 'Ð\x94миÑ\x82Ñ\x80ий',
       'Brian', 'Miguel', 'Ð\x93оÑ\x81Ñ\x82Ñ\x8c', 'Richard',
       'einem Kunden', 'roberto', 'Patrick', 'Anonymous', 'Julian',
       'Ð\x90ндÑ\x80ей', 'Mauro', 'Luca', 'Ð\x9dадежда',
       'Marco', 'Marc', 'MK', 'Angelo', 'Frank', 'Ð\x9cиÑ\x85аил',
       'Federico', 'Ð\x90Ñ\x80Ñ\x82ем', 'vincenzo', 'Matt', 'Eduardo',
       'Roberto', '???????????????? ??????????????', 'Amanda',
       'Ð\x92алеÑ\x80ий', 'Jesus', 'einer Kundin', 'Sunny',
       'patrick', 'Jessica', 'Jorge', 'xawi', 'Hugo', 'Kevin', 'David',
       'Francisco', 'Simon', 'Sam', 'Jake', 'daniel',
       'Ð\x90лекÑ\x81андÑ\x80', 'Ð\x9cаÑ\x80ина', 'Ana', 'Josh',
       'mike', 'Marcel', 'gumusoje', 'giorgio', 'E

In [45]:
df_c1['product'].unique()

array(['Apple iPhone 3GS 8GB Black Factory Unlocked / Not Jailbroken',
       'LG G2 - Smartphone libre Android (pantalla 5.2", cámara 13 Mp, 16 GB, Quad-Core 2.3 GHz, 2 GB RAM), blanco',
       'Sim Free Apple iPhone 5S 16GB Mobile Phone - Space Grey', ...,
       'Samsung Galaxy Grand Prime SM-G530H (Grey, 8GB)', 'LG L90 D410',
       'LG Optimus 4X HD P880'], dtype=object)

In [46]:
df_c1.dtypes

score      float64
author      object
product     object
dtype: object

In [47]:
df_c2 = df_c1
df_c2.head()

score           author  \
170533   10.0             John   
66599    10.0           Sergio   
41846    10.0             Nick   
33065    10.0            e-bit   
185093    8.0  Client d'Amazon   

                                                  product  
170533  Apple iPhone 3GS 8GB Black Factory Unlocked / ...  
66599   LG G2 - Smartphone libre Android (pantalla 5.2...  
41846   Sim Free Apple iPhone 5S 16GB Mobile Phone - S...  
33065           Smartphone Samsung Galaxy S7 SM-G930 32GB  
185093  Huawei - P8 Lite - Smartphone Débloqué - 4G ...

In [48]:
# Categorical boolean mask
categorical_feature_encoding = df_c2.dtypes==object
# filter categorical columns using mask and turn it into a list
categorical_cols = df_c2.columns[categorical_feature_encoding].tolist()
categorical_cols

['author', 'product']

In [49]:
# import labelencoder
from sklearn.preprocessing import LabelEncoder
# instantiate labelencoder object
le = LabelEncoder()

In [50]:
# apply le on categorical feature columns
df_c2[categorical_cols] = df_c2[categorical_cols].apply(lambda col: le.fit_transform(col))
df_c2[categorical_cols].head(10)

author  product
170533     228       96
66599      404     1190
41846      324     3394
33065      489     3476
185093     109      895
60010       45     1940
44679      343      778
226489      45      405
1775        45     2669
294209      46     1136

In [51]:
df_c2.head(50)

score  author  product
170533   10.0     228       96
66599    10.0     404     1190
41846    10.0     324     3394
33065    10.0     489     3476
185093    8.0     109      895
60010     2.0      45     1940
44679    10.0     343      778
226489    8.0      45      405
1775      8.0      45     2669
294209   10.0      46     1136
308257    6.0      45     2849
200834    8.0      45     1594
70294     6.0      45     2348
198693   10.0     110     2592
185592    4.0     109      895
289536    2.0      45      491
22704    10.0      45     1760
263773   10.0      45      233
145301    2.0     109     2594
359075   10.0      45     1802
336461   10.0     110     1774
47256    10.0      45      176
258909    8.0     377     1892
107814    8.0     652     3264
182444    6.0     302     1027
230580    2.0      46     1624
310223    2.0     349     2855
82669     4.0     328      150
47165     2.0      86      175
75189    10.0      13      708
280613    2.0      45     1246
20651     8.0      45     1516
67481    10.0     110     1048
90102     9.0     198     3736
71425     6.0      45     2349
173389    6.0      45     1028
324845    2.0      45     1766
170039    8.0     489     3460
184714    9.0     636     3721
112379    2.0      45     1542
170747   10.0     110      857
305207    8.0      89     3823
214823   10.0     318     3449
171570   10.0     632       97
261007   10.0     349     3883
369476    6.0     374     2663
312173   10.0     110      319
238148   10.0     489     3413
54197    10.0     491      279
211166   10.0     577     1360

In [52]:
df_c2.shape

(108174, 3)

In [53]:
from surprise import Dataset,Reader
from surprise.model_selection import cross_validate
from surprise import NormalPredictor
reader = Reader(rating_scale=(1, 10))
data1 = Dataset.load_from_df(df_c2[['product', 'author', 'score']], reader)

In [54]:
from surprise.model_selection import train_test_split
trainset, testset = train_test_split(data1, test_size=.25,random_state=123)

In [55]:
trainset.all_ratings()

<generator object Trainset.all_ratings at 0x000002A13896E660>

In [56]:
print(trainset.to_raw_uid(0))

1302


In [57]:
from surprise import SVD, KNNWithMeans
from surprise import accuracy

In [58]:
svd_model = SVD(n_factors=5,biased=False)
svd_model.fit(trainset)

In [59]:
testset[0]

(2027, 558, 8.0)

In [60]:
test_pred = svd_model.test(testset)

# 5. Evaluate the collaborative model. Print RMSE value. 

In [61]:
accuracy.rmse(test_pred)

RMSE: 2.9845


2.9845113385714024

In [62]:
from surprise import KNNWithMeans
from surprise import accuracy

itemb = KNNWithMeans(k=10, sim_options={ 'user_based': False})

itemb.fit(trainset)


Computing the msd similarity matrix...
Done computing similarity matrix.


- 5. Evaluate the collaborative model. Print RMSE value. 

In [63]:
test_pred=itemb.test(testset)
print(accuracy.rmse(test_pred))

RMSE: 2.7331
2.733143120437298


In [64]:
from surprise import KNNWithMeans
from surprise import accuracy

userb = KNNWithMeans(k=10, sim_options={ 'user_based': True})

userb.fit(trainset)


Computing the msd similarity matrix...
Done computing similarity matrix.


- 5. Evaluate the collaborative model. Print RMSE value. 

In [65]:
test_pred=userb.test(testset)
print(accuracy.rmse(test_pred))

RMSE: 2.7331
2.7331429492095483


- 7. Report your findings and inferences. 

- Both popularity based and collabarative filtering recommendation systems have their own benefits in providing the recommendations to the users.
- Popularity based recommendation system is non-personalised, hence it delivers the recommendation based upon the max user searches. It is not based upon the user intreset but ye is effective sometimes.
- On the other hand the collaborative filtered recommednation system is complex and more effective recommendation system that checks wide recommendations and also give imprtance over user area of interest.

- 6. Predict score (average rating) for test users. 
- 8. Try and recommend top 5 products for test users. 

In [86]:
test_pred

[Prediction(uid=2027, iid=558, r_ui=8.0, est=7.984, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=1959, iid=45, r_ui=10.0, est=6.2, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=3124, iid=491, r_ui=6.0, est=9.8, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=110, iid=346, r_ui=4.0, est=7.263209739936524, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=1513, iid=45, r_ui=8.0, est=7.4, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=2581, iid=651, r_ui=10.0, est=9.6450658138503, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=3103, iid=232, r_ui=10.0, est=8.983688842716298, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=334, iid=45, r_ui=10.0, est=5.4, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=1542, iid=45, r_ui=2.0, est=6.4, details={'actual_k': 10, 'was_impossible': False}),
 Prediction(uid=2815, iid=165, r_ui=6.0, e

In [91]:
uid = str(3465)  
iid = "489"  

# get a prediction for specific users and items.
pred = itemb.predict(uid, iid, r_ui=0.0, verbose=True)

user: 3465       item: 489        r_ui = 0.00   est = 7.83   {'was_impossible': True, 'reason': 'User and/or item is unknown.'}


In [93]:
pred = pd.DataFrame(test_pred)
pred[pred['uid'] == 3465][['iid', 'r_ui','est']].sort_values(by = 'r_ui',ascending = False).head(5)

iid  r_ui  est
12     489  10.0  9.4
13373  489  10.0  9.4
24146  489  10.0  9.4
9096   489  10.0  9.4
24058  489  10.0  9.4

- Thus above are the mentioned five top products based on the collaborative based recommendation system. Note: The recommendation is based on a particular user. Hence by user id the related recommendations can be delivered.

- 9. Try other techniques (Example: cross validation) to get better results.

In [66]:
from surprise.model_selection import cross_validate
cross_validate(svd_model, data1, measures=['RMSE', 'MAE'],cv=5, verbose=False, n_jobs=-1)

{'test_rmse': array([2.9718729 , 2.99196842, 2.96076829, 2.97874056, 2.99604717]),
 'test_mae': array([2.10274611, 2.11470147, 2.10549264, 2.10393566, 2.13108194]),
 'fit_time': (4.086010694503784,
  3.8828985691070557,
  3.9141480922698975,
  3.8829007148742676,
  2.37483549118042),
 'test_time': (0.31247901916503906,
  0.32810354232788086,
  0.34372615814208984,
  0.32810163497924805,
  0.17186331748962402)}

# 10. In what business scenario you should use popularity based Recommendation Systems ? 



- Popularity based recommendation system is related to the intrest of customers. Hence business where customers are the main priority uses popularity based recommendation system.
- Some of the examples are:
- groceries
- e commerce 
- search engines

# 11. In what business scenario you should use CF based Recommendation Systems ? 

- collaborative filtering is a method of providing similar products as recommendation to the customers.
- Business such as e commerce, amazon, netflix, spotify.
- Furthermore CF based recommendation system can be implemented in super markets so that people tend to like similar products to buy.

# 12. What other possible methods can you think of which can further improve the recommendation for different users ? 

- Recommendation systems can be improved by utilizing the advanced SVD algorithm i.e. SVD++. Also by using the deep learning algorithms. On a research basis genetic algorithms which are very powerful in computation and analysis can be employed for recommendation systems.